sklearn/dummy.py

```diff
@@ -122,7 +123,11 @@ def fit(self, X, y, sample_weight=None):
         if not self.sparse_output_:
             y = np.atleast_1d(y)
 
-        self.output_2d_ = y.ndim == 2
+        self.output_2d_ = (y.ndim == 2) and (y.shape[1] > 1)
+
+        if not self.output_2d_:
```

Comment:
Rather than resorting to custom code, can you not use `check_X_y` as I mentioned in the associated issue?

Reply:
As far as I understood it the problem with using `check_X_y` is that since multi_output needs to be true in this case, `check_X_y` will not transform y to be 1d. We can set the flag to false if the output is not 2d but this ends up as the same code as above. If I am missing something let me know.

Reply:
I am not 100% sure (and don't have time to dig into it right now) but it feels like you are using custom validation code where `check_X_y` some  arguments may need to be set properly for this use case.

The advantage of using a `check_*` function is that it is thoroughly used and tested across the code base, fixes to `check_*` will benefit you as well. By using custom code it is quite easy to introduce bugs.

Maybe what would make it easier to review is if you could summarize why `check_X_y` and/or `check_array` is not possible. Bonus points if you can add snippets because that sometimes makes it clearer than words.

Reply:
That is the snippet that demonstrates what I was saying. The problem is that since we accept multi_outputs then check_X_y does not reduce the dimension to 1d in the special case where the second dimension is just 1 thus a vector e.g. shape=(x,1).

`import numpy as np

from sklearn.utils.validation import check_X_y

X = np.array([[0], [0], [0], [0]])
y_1d = np.array([1, 2, 1, 1])
y_2d = np.array([[1], [2], [1], [1]])

output_2d_ = (y_2d.ndim == 2) 

X_new, y_new = check_X_y(X, y_2d, multi_output=output_2d_)
print(y_2d.ndim)
print(y_new.ndim)`

Reply:
Same solution to the problem exists in other parts of the code, for example in multi_layer_perceptron. 
```
915 if y.ndim == 2 and y.shape[1] == 1:
916  y = column_or_1d(y, warn=True)
```

This bug is mainly fixed in other classifiers but persists in the dummy one. One alternative would be to add a function in `validation.py` that checks whether y is multi_output and replace the repeatable custom code with that to achieve the benefits of reusability you are describing.

Reply:
OK thanks for the details! Maybe we can use `warn=True` in `column_or_1d` so that there is a warning similarly to what is done in other places?

Side-comment: to link to code inside github like you can use this [nice feature](https://help.github.com/articles/creating-a-permanent-link-to-a-code-snippet/).

Reply:
👍 Makes sense. I am uploading a new version with `warn=True`. Anything else that needed to be done in order to be ready to merge?